uv/experimental/metriccallback/impl.py

```diff
@@ -13,60 +13,55 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""The MeasurementManager class, which conducts and reports on measurements
-which may be measured at varying intervals through training."""
+"""The MetricCallback class, which conducts measurements
+which may be measured at varying times through training."""
 
 from typing import Dict, Iterable, List, Any, Union, Callable, Optional
-from uv.reporter import AbstractReporter
 
 
-class MeasurementManager:
-  """Class for a MeasurementManager.  This manager is instantiated with a
-  particular reporter, which it will use to report whatever measurements is
-  conducts.
+class MetricCallback:
+  """Class for a MetricCallback
 
-  The MeasurementManager has a set of measurements---stored in a dictionary
-  (described below)---which are to be carried out at a certain frequency, which
-  is specific to each measurement.
+  The MetricCallback has a set of measurement specifications---stored in a
+  dictionary (described below)---which are to be carried out at specified
+  training steps, which is specific to each measurement.
 
-  Whenever the MeasurementManager's process() function is called, with a
-  step number provided as an argument, the MeasurementManager decides which
-  measurements are required at each step, carries them out, and reports them.
+  Whenever the MetricCallback's measure() function is called, with a
+  step number provided as an argument, the MetricCallback decides which
+  measurements are required at each step, carries them out, and returns a
+  dictionary of the results.
 
-  In order to carry out the measurement, the Manager has a state (Dict) at each
-  step, which contains all the information necessary to perform the measurement.
-  Part of this state is static (e.g., the training dataset), while part can
-  change at each step.  This dynamic state is passed along in the process() call.
+  In order to carry out the measurement, the MetricCallback has a state (Dict)
+  at each step, which contains all the information necessary to perform the
+  measurement. Part of this state is static (e.g., the training dataset), while
+  part can change at each step.  This dynamic state is passed along in
+  the measure() call.
 
   Args:
     static_state: Dict of variables which will be used to compute the measured
                   values.  The static_state is fixed throughout time.
-    reporter: Reporter which will report the measurements.
   """
 
-  def __init__(self, static_state: Dict[str, Any], reporter: AbstractReporter):
+  def __init__(self, static_state: Dict[str, Any]):
     self.static_state = static_state
-    self.reporter = reporter
     self.measurements = {}
 
   def add_measurement(self, measurement_spec: Dict[str, Union[str, int,
```

Comment:
It might be nice to create types for `TriggerFunction` and `MeasurementFunction` using `Callable`, just to make the requirements for those explicit. Then you could use those in the function spec here as well. Just a thought, not critical at all.

Reply:
Done! I've defined
`TriggerFunction = Callable[[int], bool]`
and 
`MeasurementFunction = Callable[[Dict[str, Any]], t.Metric]`
It might be good in the future to restrict the allowed values in the state dictionary.  Currently I give complete freedom there, since this state dictionary can contain everything from a test dataset to an optimizer state, etc.